### PR TITLE
errors: expose pub getters for type-erased errors

### PR DESCRIPTION
**Pull request by muzarski**
## Motivation

Sometimes, user might want to downcast `Arc<dyn Error>` to specific error type.

Actually, there is a use case for this in cpp-rust-driver, where we need to define custom `SerializeRow` implementation. It returns a custom error type. We would like to match against it during rust-to-C error conversion - downcasting is thus required.

See: https://github.com/scylladb/cpp-rust-driver/pull/187

## Pre-review checklist

- [x] I have split my patch into logically separate commits.
- [x] All commit messages clearly explain what they change and why.
- ~[ ] I added relevant tests for new features and bug fixes.~
- [x] All commits compile, pass static checks and pass test.
- [x] PR description sums up the changes and reasons why they should be introduced.
- [x] I have provided docstrings for the public items that I want to introduce.
- ~[ ] I have adjusted the documentation in `./docs/source/`.~
- ~[ ] I added appropriate `Fixes:` annotations to PR description.~

### CODE REVIEW

File: scylla-cql/src/types/deserialize/mod.rs [hunk 1]
```diff
@@ -211,6 +211,11 @@ impl TypeCheckError {
     pub fn new(err: impl std::error::Error + Send + Sync + 'static) -> Self {
         Self(Arc::new(err))
     }
+
+    /// Retrieve an error reason.
+    pub fn get_reason(&self) -> &Arc<dyn std::error::Error + Send + Sync> {
+        &self.0
+    }
```
Comment (wprzytula) [inline scylla-cql/src/types/deserialize/mod.rs:218]: Do we prefer to expose `&Arc<dyn...>` or `&dyn...`? In the former case, it's possible to clone the underlying error; no idea if it's better or not.

Comment (Lorak-mmk) [reply scylla-cql/src/types/deserialize/mod.rs:218]: With `&dyn` it is still possible to clone the outer error so I don't see the difference.
Downcast methods on `Error` are not defined for `Arc` anyway, only for `Box`, so user will have to use methods on refs anyway.

Comment (wprzytula) [reply scylla-cql/src/types/deserialize/mod.rs:218]: Let's expose bare `&dyn` then.

Comment (muzarski) [reply scylla-cql/src/types/deserialize/mod.rs:218]: `downcast_ref` method requires `T: 'static`. See: https://doc.rust-lang.org/std/error/trait.Error.html#method.downcast_ref-1. If we return `&dyn Error`, then result will not satisfy `'static` bound. I'm curious why does it work for `Arc<dyn Error>` though - I'd be glad for an explanation.

Comment (muzarski) [reply scylla-cql/src/types/deserialize/mod.rs:218]: I tried to change `BrokenConnectionError::get_reason` accordingly, but received a compile error (because of the reasons mentioned above).

Comment (wprzytula) [reply scylla-cql/src/types/deserialize/mod.rs:218]: `Arc` is `'static`, because its lifetime is independent of any external data. It manages its data on its own, so it's `'static` just like `Vec` or `String`.

Comment (muzarski) [reply scylla-cql/src/types/deserialize/mod.rs:218]: > I tried to change `BrokenConnectionError::get_reason` accordingly, but received a compile error (because of the reasons mentioned above).

Forgot to mention that this method is used in one test - compiler error appears there

Comment (Lorak-mmk) [reply scylla-cql/src/types/deserialize/mod.rs:218]: > `downcast_ref` method requires `T: 'static`. See: https://doc.rust-lang.org/std/error/trait.Error.html#method.downcast_ref-1. If we return `&dyn Error`, then result will not satisfy `'static` bound. I'm curious why it works for `Arc<dyn Error>` though - I'd be glad for an explanation.

This is a bound on T, not on the reference. So if you downcast `&dyn Error ` to `SomeError` (so the method returns `Option<&SomeError>` then `SomeError` must be `Send + 'static` - which it typically is. What is the code and error you are getting?

Comment (muzarski) [reply scylla-cql/src/types/deserialize/mod.rs:218]: > > `downcast_ref` method requires `T: 'static`. See: https://doc.rust-lang.org/std/error/trait.Error.html#method.downcast_ref-1. If we return `&dyn Error`, then result will not satisfy `'static` bound. I'm curious why it works for `Arc<dyn Error>` though - I'd be glad for an explanation.
> 
> This is a bound on T, not on the reference.

Yeah, I just noticed that `T` is a return type. There are no lifetime bounds on arguments

Comment (muzarski) [reply scylla-cql/src/types/deserialize/mod.rs:218]: Here is a snippet from test:
```rust
        // Wait until keepaliver gots impatient and terminates router.
        // Then, the error from keepaliver will be propagated to the error receiver.
        let err = error_receiver.await.unwrap();
        let err_inner: &BrokenConnectionErrorKind = match err {
            crate::transport::connection::ConnectionError::BrokenConnection(ref e) => {
                e.get_reason().downcast_ref().unwrap()
            }
            _ => panic!("Bad error type. Expected keepalive timeout."),
        };
        assert_matches!(err_inner, BrokenConnectionErrorKind::KeepaliveTimeout(_));
```

It works when `BrokenConnectionError` returns `Arc<dyn ...>`, but when I change the return type to `&dyn Error`, I get the following error:
```
error[E0597]: `err.0` does not live long enough
    --> scylla/src/transport/connection.rs:2782:77
     |
2782 |             crate::transport::connection::ConnectionError::BrokenConnection(ref e) => {
     |                                                                             ^^^^^ borrowed value does not live long enough
2783 |                 e.get_reason().downcast_ref().unwrap()
     |                 -------------- argument requires that `err.0` is borrowed for `'static`
...
2796 |     }
     |     - `err.0` dropped here while still borrowed
```

Comment (Lorak-mmk) [reply scylla-cql/src/types/deserialize/mod.rs:218]: What does it say if you add `drop(err);` after `assert_matches`?

Comment (muzarski) [reply scylla-cql/src/types/deserialize/mod.rs:218]: > What does it say if you add `drop(err);` after `assert_matches`?

"cannot move out of `err` because it is borrowed"

Comment (Lorak-mmk) [reply scylla-cql/src/types/deserialize/mod.rs:218]: Can you try to assign `e` to a variable first? So the snippet would look like this:
```
        // Wait until keepaliver gots impatient and terminates router.
        // Then, the error from keepaliver will be propagated to the error receiver.
        let err = error_receiver.await.unwrap();
        let err_inner = match err {
            crate::transport::connection::ConnectionError::BrokenConnection(ref e) => {
                e
            }
            _ => panic!("Bad error type. Expected keepalive timeout."),
        };
        let err_typed: &BrokenConnectionErrorKind = err_inner.get_reason().downcast_ref().unwrap();
        assert_matches!(err_typed, BrokenConnectionErrorKind::KeepaliveTimeout(_));
```

Comment (muzarski) [reply scylla-cql/src/types/deserialize/mod.rs:218]: The same error:
```
error[E0597]: `err.0` does not live long enough
    --> scylla/src/transport/connection.rs:2782:77
     |
2782 |             crate::transport::connection::ConnectionError::BrokenConnection(ref e) =...
     |                                                                             ^^^^^ borrowed value does not live long enough
...
2785 |         let err_typed: &BrokenConnectionErrorKind = err_inner.get_reason().downcast_...
     |                                                     ---------------------- argument requires that `err.0` is borrowed for `'static`
...
2795 |     }
     |     - `err.0` dropped here while still borrowed
```

Comment (Lorak-mmk) [reply scylla-cql/src/types/deserialize/mod.rs:218]: Wtf. 
What docs.rs show:
image
What is really there when you click "source":

Comment (Lorak-mmk) [reply scylla-cql/src/types/deserialize/mod.rs:218]: Wtf. 
What docs.rs show:
image
What is really there when you click "source":
image

Comment (Lorak-mmk) [reply scylla-cql/src/types/deserialize/mod.rs:218]: In that case, I think the best solution is to create our own downcast_ref method on TypeCheckError etc - that should be possible because inside we have access to Arc, right?

Comment (muzarski) [reply scylla-cql/src/types/deserialize/mod.rs:218]: I'll try it out. But I still don't really get it why it works for `Arc<dyn Error>`, while it does not for `&dyn Error`. AFAIU, what happens when we call `downcast_ref` on `Arc<dyn Error>` is value inside `Arc` is implicitly dereferenced first (i.e. `&dyn Error` is retrieved automatically via `Deref` trait). Then, `downcast_ref` is called on obtained `&dyn Error` reference. It seems almost identical to me.

**Comment by muzarski** [comment]
v2: replaced public getters with custom `downcast_ref` implementations for type-erased error types.

**Review by wprzytula — APPROVED** [review]

File: scylla-cql/src/types/deserialize/mod.rs (hunk 1 above)
Comment (Lorak-mmk) [reply scylla-cql/src/types/deserialize/mod.rs]: We should understand it and decide on how to best fix that before merging. I'll try to look into this today or during the weekend.

**Review by Lorak-mmk — APPROVED** [review]
Maybe not very related: Error impls / deriving seem to be a bit inconsistent:
- SerializationError / DeserializationError has derive(Error) + impl Display
- BrokenConnectionError additionally has `#[error("Connection broken, reason: {0}")]` and no manual Display impl
- TypeCheckError has `#[error(transparent)]` and no manual Display impl

I wonder if there is a good reason for those differences.

**Comment by wprzytula** [comment]
> Maybe not very related: Error impls / deriving seem to be a bit inconsistent:
> 
>     * SerializationError / DeserializationError has derive(Error) + impl Display
> 
>     * BrokenConnectionError additionally has `#[error("Connection broken, reason: {0}")]` and no manual Display impl
> 
>     * TypeCheckError has `#[error(transparent)]` and no manual Display impl
> 
> 
> I wonder if there is a good reason for those differences.

`TypeCheckError` should be consistent with `DeserializationError`. If it's not, let's fix this.

**Review by Lorak-mmk — DISMISSED** [review]

**Comment by Lorak-mmk** [comment]
Sorry, one more thing: `TypeCheckError` is the only one of those 4 types that has its inside `pub(crate) ` instead of private. Can we change that?

**Comment by muzarski** [comment]
> Sorry, one more thing: `TypeCheckError` is the only one of those 4 types that has its inside `pub(crate) ` instead of private. Can we change that? I doubt that with `downcast_ref` there is need to use this member directly by other places in the code.

This member is actually used directly in `scylla-cql/src/frame/response/result.rs` (on line 1052)

File: scylla/src/transport/errors.rs [hunk 2]
```diff
@@ -734,6 +733,12 @@ impl BrokenConnectionError {
     }
 }
 
+impl std::fmt::Display for BrokenConnectionError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "Connection broken, reason: {}", self.0)
+    }
+}
+
```
Comment (wprzytula) [inline scylla/src/transport/errors.rs:741]: Hmm, why this change from auto-generated Display to boilerplate'ish Display impl? Why not the other way round, so that all three error types use `#[error()]` shorthand?

Comment (muzarski) [reply scylla/src/transport/errors.rs:741]: True.

**Review by wprzytula — APPROVED** [review]

**Comment by muzarski** [comment]
Flaky test `transport::connection::tests::test_coalescing` failed yet again...

**Comment by wprzytula** [comment]
> Flaky test transport::connection::tests::test_coalescing failed yet again...

Another time in a row.

**Comment by Lorak-mmk** [comment]
Well, flakiness is one more reason to change this test

**Review by Lorak-mmk — APPROVED** [review]
